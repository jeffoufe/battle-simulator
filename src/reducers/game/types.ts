import {
    PLAYER_ROLL,
    MONSTER_ROLL,
    ALL_ROLLS_DONE,
    RESET_TURN,
    COMPUTE_TURN,
    RESET_GAME
} from './constants';

export interface PlayerRollAction {
    type: typeof PLAYER_ROLL
}


export interface MonsterRollAction {
    type: typeof MONSTER_ROLL,
    payload: { result: number }
}

export interface AllRollsDoneAction {
    type: typeof ALL_ROLLS_DONE,
    payload: { result: number }
}

export interface ResetTurnAction {
    type: typeof RESET_TURN
}

export interface ComputeTurnAction {
    type: typeof COMPUTE_TURN
}

export interface ResetGameAction {
    type: typeof RESET_GAME
}

export type GameReducerAction =
    PlayerRollAction |
    MonsterRollAction |
    AllRollsDoneAction |
    ResetTurnAction |
    ComputeTurnAction |
    ResetGameAction