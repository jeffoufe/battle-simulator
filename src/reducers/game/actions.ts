import { 
    PLAYER_ROLL, 
    RESET_GAME,
    MONSTER_ROLL,
    END_TURN
} from './constants';

export const clickAttack = () => ({
    type: PLAYER_ROLL
})

export const resetGame = () => ({
    type: RESET_GAME
})

export const monsterRoll = (result: number) => ({
    type: MONSTER_ROLL,
    payload: { result }
})

export const endTurn = (result: number) => ({
    type: END_TURN,
    payload: { result }
})