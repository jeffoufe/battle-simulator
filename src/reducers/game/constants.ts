export const PLAYER_ROLL = 'GAME/PLAYER_ROLL';
export const MONSTER_ROLL = 'GAME/MONSTER_ROLL';
export const ALL_ROLLS_DONE = 'GAME/ALL_ROLLS_DONE';
export const RESET_TURN = 'GAME/RESET_TURN';
export const END_TURN = 'GAME/END_TURN';
export const COMPUTE_TURN = 'GAME/COMPUTE_TURN';
export const RESET_GAME = 'GAME/RESET_GAME';