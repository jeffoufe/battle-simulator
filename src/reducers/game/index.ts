import { State } from '..';
import { GameReducerAction } from './types';
import {
  PLAYER_ROLL,
  MONSTER_ROLL,
  ALL_ROLLS_DONE,
  RESET_TURN,
  COMPUTE_TURN,
  RESET_GAME
} from './constants';

export const CHARACTER_INITIAL_STATE = {
  health: 100,
  result: 0,
  isRolling: false
};

export const INITIAL_STATE = {
  isTurnActive: false,
  player: CHARACTER_INITIAL_STATE,
  monster: CHARACTER_INITIAL_STATE
}

export default (state: State = INITIAL_STATE, action: GameReducerAction) => {
  switch (action.type) {
    case PLAYER_ROLL:
      return {
        ...state,
        isTurnActive: true,
        player: {
          ...state.player,
          isRolling: true
        }
      }
    case MONSTER_ROLL:
      return {
        ...state,
        monster: {
          ...state.monster,
          isRolling: true
        },
        player: {
          ...state.player,
          isRolling: false,
          result: action.payload.result
        }
      }
    case ALL_ROLLS_DONE:
      return {
        ...state,
        monster: {
          ...state.monster,
          isRolling: false,
          result: action.payload.result
        }
      }
    case COMPUTE_TURN:
      const damage = state.player.result - state.monster.result;
      const character = damage < 0 ? 'player' : 'monster';
      return {
        ...state,
        [character]: {
          ...state[character],
          health: state[character].health - (Math.abs(damage))
        }
      }
    case RESET_TURN:
      return {
        ...state,
        isTurnActive: false,
        player: {
          ...state.player,
          result: 0
        },
        monster: {
          ...state.monster,
          result: 0
        }
      }
    case RESET_GAME:
      return INITIAL_STATE;
    default:
      return state;
  }
};