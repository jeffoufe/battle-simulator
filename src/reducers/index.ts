export { default as gameReducer } from './game';

interface Character {
    isRolling: boolean,
    health: number,
    result: number
};

export interface State {
    player: Character,
    monster: Character,
    isTurnActive: boolean
}