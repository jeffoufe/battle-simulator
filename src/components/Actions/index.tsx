import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { clickAttack, resetGame } from '../../reducers/game/actions';
import {
    ActionsContainer,
    AttackButton,
    ActionsInfo,
    GameOver,
    Victory
} from './components';

const Actions = () => {
    const dispatch = useDispatch();
    const { isTurnActive, player, monster } = useSelector((state: any) => state);
    const playerResult = player.result;
    const monsterResult = monster.result;
    const isGameFinished = player.health <= 0 || monster.health <= 0;
    const damage = player.result - monster.result;

    const onClickAttack = () => dispatch(clickAttack());
    const onResetGame = () => dispatch(resetGame());

    return (
        <ActionsContainer>
            {player.health <= 0 && (
                <GameOver>
                    Game Over
                </GameOver>
            )}
            {monster.health <= 0 && (
                <Victory>
                    You Win
                </Victory>
            )}
            {!!(playerResult && monsterResult) && (
                <ActionsInfo>
                    {!!(damage)
                        ? `You ${damage < 0 ? 'Are Hit' : 'Hit'} For ${Math.abs(damage)}`
                        : 'Draw Turn'
                    }
                </ActionsInfo>
            )}
            {!isTurnActive && (
                <AttackButton onClick={isGameFinished ? onResetGame : onClickAttack}>
                    {isGameFinished ? 'Reset': 'Attack!'}
                </AttackButton>
            )}
        </ActionsContainer>
    );
}

Actions.displayName = 'Actions';
export default Actions;