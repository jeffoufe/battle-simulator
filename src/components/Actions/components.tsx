import styled from 'styled-components';
import { breakpoints } from '../../constants';

export const ActionsContainer = styled.div`
    width: 100%;
    text-align: center;
    display: flex;
    flex-direction: column;
`;

export const ActionsInfo = styled.p`
    font-size: 1em;
    text-transform: uppercase;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        font-size: 1.5em;
    }
`

export const GameOver = styled(ActionsInfo)`
    color: red;
`

export const Victory = styled(ActionsInfo)`
    color: green;
`

export const AttackButton = styled.button`
    color: white;
    background-color: red;
    padding: 10px 20px;
    border: 0px;
    margin-bottom: 20px;
    width: 50%;
    margin-left: 25%;
    font-size: 1em;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        font-size: 1.5em;
    }
`

ActionsContainer.displayName = 'ActionsContainer';
ActionsInfo.displayName = 'ActionsInfo';
GameOver.displayName = 'GameOver';
Victory.displayName = 'Victory';
AttackButton.displayName = 'AttackButton';