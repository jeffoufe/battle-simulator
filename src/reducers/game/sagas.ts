import { ALL_ROLLS_DONE, END_TURN, RESET_TURN, COMPUTE_TURN } from './constants';
import { takeEvery, put, select, delay } from 'redux-saga/effects';

interface EndTurnAction {
    payload: {
        result: number
    }
}

function* endTurn(action: EndTurnAction) {
    const { result } = action.payload;
    yield put({
        type: ALL_ROLLS_DONE,
        payload: { result }
    })
    const playerResult = select((state: any) => state.player.result);
    const monsterResult = select((state: any) => state.monster.result);
    if (playerResult === monsterResult) {
        yield delay(1000);
        yield put({
            type: RESET_TURN
        });
    } else {
        const order = playerResult < monsterResult;
        yield put({
            type: COMPUTE_TURN,
            payload: { order }
        })
        yield delay(1000);
        yield put({
            type: RESET_TURN
        });
    }
}   

export function* watchEndTurn() {
    // @ts-ignore
    yield takeEvery(END_TURN, endTurn);
}