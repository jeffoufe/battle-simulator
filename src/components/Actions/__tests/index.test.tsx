import React from 'react';
import { shallow } from 'enzyme';
import Actions from '..';
import { INITIAL_STATE } from '../../../reducers/game';
import { clickAttack, resetGame } from '../../../reducers/game/actions';
import * as redux from "react-redux";

const mockDispatchFn = jest.fn()

describe('Actions', () => {
    beforeEach(() => {
        const useDispatchSpy = jest.spyOn(redux, 'useDispatch'); 
        useDispatchSpy.mockReturnValue(mockDispatchFn);
    })

    it('Game start', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue(INITIAL_STATE);

        const wrapper = shallow(
            <Actions />
        );
        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GameOver').length).toEqual(0);
        expect(wrapper.find('Victory').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').length).toEqual(0);
        expect(wrapper.find('AttackButton').length).toEqual(1);

        wrapper.find('AttackButton').prop('onClick')();
        expect(mockDispatchFn).toHaveBeenCalledWith(clickAttack());
    })

    it('Dices drawn - Player Hits', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({
            isTurnActive: true,
            player: {
                ...INITIAL_STATE.player,
                result: 6
            },
            monster: {
                ...INITIAL_STATE.monster,
                result: 3
            }
        });

        const wrapper = shallow(
            <Actions />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GameOver').length).toEqual(0);
        expect(wrapper.find('Victory').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').length).toEqual(1);
        expect(wrapper.find('AttackButton').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').text()).toEqual('You Hit For 3');
    })

    it('Dices drawn - Monster Hits', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({
            isTurnActive: true,
            player: {
                ...INITIAL_STATE.player,
                result: 5
            },
            monster: {
                ...INITIAL_STATE.monster,
                result: 6
            }
        });

        const wrapper = shallow(
            <Actions />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GameOver').length).toEqual(0);
        expect(wrapper.find('Victory').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').length).toEqual(1);
        expect(wrapper.find('AttackButton').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').text()).toEqual('You Are Hit For 1');
    })

    it('Dices drawn - Monster Hits', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({
            isTurnActive: true,
            player: {
                ...INITIAL_STATE.player,
                result: 5
            },
            monster: {
                ...INITIAL_STATE.monster,
                result: 5
            }
        });

        const wrapper = shallow(
            <Actions />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GameOver').length).toEqual(0);
        expect(wrapper.find('Victory').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').length).toEqual(1);
        expect(wrapper.find('AttackButton').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').text()).toEqual('Draw Turn');
    })

    it('Game Over', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({
            ...INITIAL_STATE,
            player: {
                ...INITIAL_STATE.player,
                health: 0
            }
        });

        const wrapper = shallow(
            <Actions />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GameOver').length).toEqual(1);
        expect(wrapper.find('Victory').length).toEqual(0);
        expect(wrapper.find('ActionsInfo').length).toEqual(0);
        expect(wrapper.find('AttackButton').length).toEqual(1);
        wrapper.find('AttackButton').prop('onClick')();
        expect(mockDispatchFn).toHaveBeenCalledWith(resetGame());
    })

    it('Victory', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({
            ...INITIAL_STATE,
            monster: {
                ...INITIAL_STATE.monster,
                health: 0
            }
        });

        const wrapper = shallow(
            <Actions />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GameOver').length).toEqual(0);
        expect(wrapper.find('Victory').length).toEqual(1);
        expect(wrapper.find('ActionsInfo').length).toEqual(0);
        expect(wrapper.find('AttackButton').length).toEqual(1);
        wrapper.find('AttackButton').prop('onClick')();
        expect(mockDispatchFn).toHaveBeenCalledWith(resetGame());
    })
})

