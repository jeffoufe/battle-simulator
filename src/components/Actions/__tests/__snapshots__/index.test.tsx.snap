// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions Dices drawn - Monster Hits 1`] = `
<ActionsContainer>
  <ActionsInfo>
    You Are Hit For 1
  </ActionsInfo>
</ActionsContainer>
`;

exports[`Actions Dices drawn - Monster Hits 2`] = `
<ActionsContainer>
  <ActionsInfo>
    Draw Turn
  </ActionsInfo>
</ActionsContainer>
`;

exports[`Actions Dices drawn - Player Hits 1`] = `
<ActionsContainer>
  <ActionsInfo>
    You Hit For 3
  </ActionsInfo>
</ActionsContainer>
`;

exports[`Actions Game Over 1`] = `
<ActionsContainer>
  <GameOver>
    Game Over
  </GameOver>
  <AttackButton
    onClick={[Function]}
  >
    Reset
  </AttackButton>
</ActionsContainer>
`;

exports[`Actions Game start 1`] = `
<ActionsContainer>
  <AttackButton
    onClick={[Function]}
  >
    Attack!
  </AttackButton>
</ActionsContainer>
`;

exports[`Actions Victory 1`] = `
<ActionsContainer>
  <Victory>
    You Win
  </Victory>
  <AttackButton
    onClick={[Function]}
  >
    Reset
  </AttackButton>
</ActionsContainer>
`;
